src/IRM/IRMAdaptiveLinearKink.sol

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {IIRM} from "evk/InterestRateModels/IIRM.sol";
+import {ExpWad} from "./lib/ExpWad.sol";
+
+/// @title IRMAdaptiveLinearKink
+/// @custom:contact security@euler.xyz
+/// @author Euler Labs (https://www.eulerlabs.com/).
+/// @author Adapted from Morpho Labs (https://github.com/morpho-org/morpho-blue-irm/).
+/// @notice A Linear Kink IRM with an adaptive mechanism based on exponential growth/decay.
+/// As utilization persists above/below the kink the Linear Kink IRM is translated up/down.
+/// @dev The model is parameterized by `(kink, kinkRate, slope)` instead of `(kink, baseRate, slope1, slope2)`.
+/// The `slope` parameter is equivalent to `slope2` (above kink), whereas `slope1` (under kink) is `1/slope`.
+/// The `kinkRate` parameter is the adaptive component in this model.
+contract IRMAdaptiveLinearKink is IIRM {
+    /// @dev Unit for internal precision.
+    int256 internal constant WAD = 1e18;
+    /// @notice The utilization rate targeted by the model.
+    /// @dev In WAD units.
+    int256 public immutable kink;
+    /// @notice The initial interest rate when utilization is at kink.
+    /// @dev In WAD per second units.
+    /// When the IRM is initialized for a vault this is the rate at kink that is assigned.
+    int256 public immutable initialKinkRate;
+    /// @notice The minimum interest rate when utilization is at kink that the model can adjust to.
+    /// @dev In WAD per second units.
+    int256 public immutable minKinkRate;
+    /// @notice The maximum interest rate when utilization is at kink that the model can adjust to.
+    /// @dev In WAD per second units.
+    int256 public immutable maxKinkRate;
+    /// @notice The steepness of interest rate line above the kink. The line below the kink has the inverse slope.
+    /// @dev In WAD per second units. The line below the kink has the inverse slope.
+    int256 public immutable slope;
+    /// @notice The speed at which the rate at kink is adjusted up or down.
+    /// @dev In WAD per second units. For example, an adjustment speed of `2e18 / 24 hours` will make the model
+    /// double `kinkRate` if the rate remains at 100% for a day.
+    int256 public immutable adjustmentSpeed;
+
+    /// @notice Cached state of the interest rate model.
+    struct IRState {
+        /// @dev The current rate at kink.
+        int208 kinkRate;
+        /// @dev The timestamp of the last update to the model.
+        uint48 lastUpdate;
+    }
+
+    /// @notice Get the cached state of a vault's irm.
+    /// @return kinkRate The last computed rate at kink.
+    /// @return lastUpdate The last update timestamp.
+    /// @dev Note that this state may be outdated. Use `computeInterestRateView` for the latest interest rate.
```

Comment:
If stored value is outdated, maybe make the mapping internal?  (That's what EVault is doing as well).

Reply:
Maybe. But `kinkRate` (and `fullRate` in the other IRM) is useful for Lens/UI to be able to graph the interest rate model. If it were internal then Lens would have to replicate the calculation math, which is messy. Maybe we could expose a `computeKinkRateView`?

Reply:
Exactly, but looks like computeKinkRateView is external already. Lens should just use that, internal state is just an input to the calculation, has no practical impact.

Reply:
I made `irState` internal and added a new function `compute[Kink/Full]RateView`. I would love to combine this one with the IIRM's `computeInterestRateView` but unfortunately that would break the interface.